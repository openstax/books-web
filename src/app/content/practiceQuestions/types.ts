import { LinkedArchiveTree, LinkedArchiveTreeSection } from '../types';

export interface State {
  isEnabled: boolean;
  summary: PracticeQuestionsSummary | null;
  open: boolean;
  selectedSection: LinkedArchiveTreeSection | null;
  currentQuestionIndex: number | null;
  questions: PracticeQuestion[];
}

export interface PracticeQuestionsSummary {
  countsPerSource: {
    [key: string]: number;
  };
}

interface PracticeAnswer {
  id: number;
  content_html: string;
  correctness: '0.0' | '1.0';
  feedback_html: string;
}

export interface PracticeQuestion {
  group_uuid: string;
  uid: string;
  stem_html: string;
  answers: PracticeAnswer[];
}

export type PracticeQuestions = PracticeQuestion[];

export type PracticeQuestionsLocationFilters = Map<
  string,
  { section: LinkedArchiveTree, children: LinkedArchiveTreeSection[] }
>;
