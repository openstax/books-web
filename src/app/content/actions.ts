import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as searchActions from './search/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();

export const closeCallToActionPopup = createStandardAction('Content/closeCallToActionPopup')();
export const openCallToActionPopup = createStandardAction('Content/openCallToActionPopup')();
