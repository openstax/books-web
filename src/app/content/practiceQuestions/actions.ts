import { createStandardAction } from 'typesafe-actions';
import { PracticeQuestionsSummary } from './types';

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();
