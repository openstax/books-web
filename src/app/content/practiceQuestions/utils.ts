import { isLinkedArchiveTreeSection } from '../guards';
import { Book, LinkedArchiveTreeSection } from '../types';
import { findArchiveTreeNodeById, flattenArchiveTree } from '../utils/archiveTreeUtils';
import { stripIdVersion } from '../utils/idUtils';
import { PracticeQuestionsSummary } from './types';

export const pageHasPracticeQuestions = (pageId: string, summary: PracticeQuestionsSummary) => {
  return Boolean(summary.countsPerSource[pageId]);
};

export const getPracticeQuestionsLocationFilters = (
  summary: PracticeQuestionsSummary | null, book: Book | undefined
) => {
  // key is an id of a parent of sections stored in a value
  const locationFilters: Map<string, LinkedArchiveTreeSection[]> = new Map();

  if (!book || !summary) { return locationFilters; }

  const tree = flattenArchiveTree(book.tree);

  for (const node of tree) {
    if (isLinkedArchiveTreeSection(node) && summary.countsPerSource[stripIdVersion(node.id)]) {
      const parentId = stripIdVersion(node.parent.id);
      locationFilters.set(parentId, [...locationFilters.get(parentId) || [], node]);
    }
  }

  return locationFilters;
};

const flattenLocationFilters = (locationFilters: Map<string, LinkedArchiveTreeSection[]>) => {
  let flattened: LinkedArchiveTreeSection[] = [];
  for (const sections of locationFilters.values()) {
    flattened = flattened.concat(sections);
  }
  return flattened;
};

export const getNextPageWithPracticeQuestions = (
  nodeId: string, locationFilters: Map<string, LinkedArchiveTreeSection[]>, book: Book | undefined
) => {
  if (!book) { return; }

  const node = findArchiveTreeNodeById(book.tree, nodeId);

  if (!node) { return; }

  const flatLocationFilters = flattenLocationFilters(locationFilters);
  const flatTree = flattenArchiveTree(book.tree);
  const nodeIndex = flatTree.findIndex((search) => search.id === node.id);
  const remainingNodes = flatTree.slice(nodeIndex + 1).map((section) => stripIdVersion(section.id));

  return flatLocationFilters.find((filter) => remainingNodes.includes(filter.id));
};
