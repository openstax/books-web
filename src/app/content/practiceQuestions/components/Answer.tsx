import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { textRegularStyle } from '../../../components/Typography';
import theme from '../../../theme';
import ContentExcerpt from '../../components/ContentExcerpt';
import { LinkedArchiveTreeSection } from '../../types';
import { PracticeAnswer, PracticeQuestionStyles } from '../types';
import contants from './contants';

interface AnswerProps {
  isSelected: boolean;
  isSubmitted: boolean;
  showCorrect: boolean;
  answer: PracticeAnswer;
  choiceIndicator: string;
  onSelect: () => void;
  source: LinkedArchiveTreeSection;
}

// tslint:disable-next-line: variable-name
const AnswerExcerpt = styled.span`
  ${textRegularStyle}
  width: 100%;
  padding: 0;
`;

// tslint:disable-next-line: variable-name
const AnswerAlignment = styled.div`
  display: flex;
  align-items: center;
  min-height: 4rem;
`;

// tslint:disable-next-line: variable-name
const AnswerContent = styled.div`
  margin-left: 1.6rem;
`;

// tslint:disable-next-line: variable-name
const AnswerLabel = styled.label`
  background-color: ${(props: {style: PracticeQuestionStyles}) => props.style.focused};
  border: 1.5px solid ${(props: {style: PracticeQuestionStyles}) => props.style.passive};
  min-width: 4rem;
  min-height: 4rem;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  color: ${(props: {style: PracticeQuestionStyles}) => props.style.fontColor};

  input {
    position: absolute;
    opacity: 0;
    height: 0;
    width: 0;
  }
`;

// tslint:disable-next-line: variable-name
export const AnswerBlock = styled.div`
  padding: 0 2rem 2.4rem;
  display: flex;
  align-items: flex-start;
  cursor: pointer;

  &:hover {
    ${AnswerLabel} {
      border-color: ${(props: {style: PracticeQuestionStyles}) => props.style.hovered};
    }
  }

  ${theme.breakpoints.mobile(css`
    padding: 0 1rem 2.4rem;

    ${ContentExcerpt} {
      padding: 0;
    }
  `)}
`;

// tslint:disable-next-line: variable-name
const Answer = (props: AnswerProps) => {
  const answerStyle = props.isSelected ? contants.selectedAnswerStyle : contants.unselectedAnswerStyle;

  return <AnswerBlock style={answerStyle} onClick={props.onSelect}>
    <FormattedMessage
      id='i18n:practice-questions:popup:answers:choice'
      values={{choiceIndicator: props.choiceIndicator.toUpperCase()}}
    >{(msg: string) =>
      <AnswerLabel style={answerStyle} htmlFor={props.answer.id} aria-label={msg}>
        {props.choiceIndicator}
        <input
          type='radio'
          id={`${props.answer.id}`}
          name={props.choiceIndicator}
          checked={props.isSelected}
          onChange={props.onSelect}
        />
      </AnswerLabel>
    }</FormattedMessage>
    <AnswerAlignment>
      <AnswerContent>
        <AnswerExcerpt>{props.answer.content_html}</AnswerExcerpt>
      </AnswerContent>
    </AnswerAlignment>
  </AnswerBlock>;
};

export default Answer;
