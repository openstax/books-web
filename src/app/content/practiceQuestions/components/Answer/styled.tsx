import flow from 'lodash/fp/flow';
import styled, { css } from 'styled-components/macro';
import { textRegularStyle } from '../../../../components/Typography';
import theme from '../../../../theme';
import ContentExcerpt from '../../../components/ContentExcerpt';

// tslint:disable-next-line: variable-name
export const AnswerExcerpt = styled.span`
  ${textRegularStyle}
  width: 100%;
  padding: 0;
  overflow: auto;

  * {
    overflow: initial;
  }
`;

// tslint:disable-next-line: variable-name
export const AnswerAlignment = styled.div`
  display: flex;
  align-items: center;
  min-height: 4rem;
  flex: 1;
`;

// tslint:disable-next-line: variable-name
export const AnswerContent = styled.div`
  margin-left: 1.6rem;
  overflow: initial;
`;

// tslint:disable-next-line: variable-name
export const AnswerLabel = styled.label`
  min-width: 4rem;
  min-height: 4rem;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;

  input {
    position: absolute;
    opacity: 0;
    height: 0;
    width: 0;
  }
`;

const answerThemes = {
  correct: {
    background: '#77AF42',
    border: '#77AF42',
    borderHovered: '#77AF42',
    fontColor: theme.color.neutral.base,
    fontColorActive: '#77AF42',
  },
  incorrect: {
    background: theme.color.primary.red.base,
    border: theme.color.primary.red.base,
    borderHovered: theme.color.primary.red.base,
    fontColor: theme.color.neutral.base,
    fontColorActive: theme.color.primary.red.base,
  },
  selected: {
    background: theme.color.secondary.lightBlue.base,
    border: theme.color.secondary.lightBlue.base,
    borderHovered: theme.color.secondary.lightBlue.base,
    fontColor: theme.color.neutral.base,
    fontColorActive: theme.color.secondary.lightBlue.base,
  },
  unselected: {
    background: theme.color.neutral.base,
    border: '#C6C6C6',
    borderHovered: theme.color.secondary.lightBlue.base,
    fontColor: '#606163',
    fontColorActive: '#C6C6C6',
  },
};

interface AnswerBlockProps {
  showCorrect: boolean;
  isCorrect: boolean;
  isSubmitted: boolean;
  isSelected: boolean;
}

const getAnswerTheme = (props: AnswerBlockProps) => {
  if (
    (props.showCorrect && props.isCorrect) || (props.isSubmitted && props.isSelected)
  ) {
    return props.isCorrect ? answerThemes.correct : answerThemes.incorrect;
  } else {
    return props.isSelected ? answerThemes.selected : answerThemes.unselected;
  }
};

const getAnswerThemeCss = (answerTheme: typeof answerThemes[keyof typeof answerThemes]) => css`
  ${AnswerLabel} {
    color: ${answerTheme.fontColor};
    background-color: ${answerTheme.background};
    border: 1.5px solid ${answerTheme.border};
  }

  &:hover {
    ${AnswerLabel} {
      border-color: ${answerTheme.borderHovered};
    }
  }
`;

// tslint:disable-next-line: variable-name
export const AnswerBlock = styled.div`
  padding: 0 2rem 2.4rem;
  display: flex;
  align-items: flex-start;
  cursor: ${(props: AnswerBlockProps) => props.isSubmitted ? 'not-allowed' : 'pointer'};
  ${flow(getAnswerTheme, getAnswerThemeCss)}
  ${theme.breakpoints.mobile(css`
    padding: 0 1rem 2.4rem;

    ${ContentExcerpt} {
      padding: 0;
    }
  `)}
`;

// tslint:disable-next-line: variable-name
export const StyledAnswerResult = styled.div`
  ${textRegularStyle}
  color: ${({isCorrect}: {isCorrect: boolean}) => answerThemes[isCorrect ? 'correct' : 'incorrect'].fontColorActive};
`;
