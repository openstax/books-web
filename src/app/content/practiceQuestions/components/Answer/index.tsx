import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { LinkedArchiveTreeSection } from '../../../types';
import { PracticeAnswer } from '../../types';
import { AnswerAlignment, AnswerBlock, AnswerContent, AnswerExcerpt, AnswerLabel, StyledAnswerResult } from './styled';

interface AnswerResultProps {
  showCorrect: boolean;
  isSubmitted: boolean;
  isSelected: boolean;
  isCorrect: boolean;
}

// tslint:disable-next-line: variable-name
const AnswerResult = ({ showCorrect, isSelected, isSubmitted, isCorrect }: AnswerResultProps) => {
  const resultMsgKey = isCorrect
    ? 'i18n:practice-questions:popup:correct'
    : 'i18n:practice-questions:popup:incorrect';

  if (
    !(isSubmitted && isSelected)
    && !(isSubmitted && showCorrect && isCorrect)
  ) {
    return null;
  }

  return <StyledAnswerResult isCorrect={isCorrect}>
    <FormattedMessage id={resultMsgKey}>
      {(msg: string) => msg}
    </FormattedMessage>
  </StyledAnswerResult>;
};

interface AnswerProps {
  isSelected: boolean;
  isSubmitted: boolean;
  showCorrect: boolean;
  answer: PracticeAnswer;
  choiceIndicator: string;
  onSelect: () => void;
  source: LinkedArchiveTreeSection;
}

// tslint:disable-next-line: variable-name
const Answer = ({
  answer,
  showCorrect,
  choiceIndicator,
  isSubmitted,
  isSelected,
  onSelect,
}: AnswerProps) => {
  const answerRef = React.useRef<HTMLElement>(null);
  const isCorrect = answer.correctness === '1.0';

  React.useEffect(() => {
    if (showCorrect && isCorrect && answerRef.current) {
      answerRef.current.focus();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [showCorrect, isCorrect]);

  return <AnswerBlock
    showCorrect={showCorrect}
    isCorrect={isCorrect}
    isSubmitted={isSubmitted}
    isSelected={isSelected}
    onClick={onSelect}
  >
    <FormattedMessage
      id='i18n:practice-questions:popup:answers:choice'
      values={{choiceIndicator: choiceIndicator.toUpperCase()}}
    >{(msg: string) =>
      <React.Fragment>
        <input
          id={choiceIndicator}
          type='radio'
          name={choiceIndicator}
          checked={isSelected}
          disabled={isSubmitted}
          onChange={onSelect}
        />
        <AnswerLabel ref={answerRef} aria-label={msg} for={choiceIndicator} tabIndex={-1}>
          {choiceIndicator}
        </AnswerLabel>
      </React.Fragment>
    }</FormattedMessage>
    <AnswerAlignment>
      <AnswerContent tabIndex={0}>
        <AnswerExcerpt>{answer.content_html}</AnswerExcerpt>
        <AnswerResult
          showCorrect={showCorrect}
          isSelected={isSelected}
          isSubmitted={isSubmitted}
          isCorrect={isCorrect}
        />
      </AnswerContent>
    </AnswerAlignment>
  </AnswerBlock>;
};

export default Answer;
