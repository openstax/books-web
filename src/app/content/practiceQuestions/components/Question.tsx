import React from 'react';
import styled from 'styled-components/macro';
import { h4Style } from '../../../components/Typography';
import theme from '../../../theme';
import ContentExcerpt from '../../components/ContentExcerpt';
import { LinkedArchiveTreeSection } from '../../types';
import { PracticeAnswer, PracticeQuestion } from '../types';

interface QuestionProps {
  question: PracticeQuestion;
  isSubmitted?: boolean;
  showCorrect?: boolean;
  source: LinkedArchiveTreeSection;
  onSelectAnswer?: (answer: PracticeAnswer) => void;
}

// tslint:disable-next-line: variable-name
const QuestionWrapper = styled.div`
  padding: 0 ${theme.padding.page.desktop}rem;
`;

// tslint:disable-next-line: variable-name
const QuestionContent = styled(ContentExcerpt)`
  ${h4Style}
  font-weight: bold;
  color: ${theme.color.primary.gray.base};
`;

// tslint:disable-next-line: variable-name
const Question = (props: QuestionProps) => {

  return <QuestionWrapper>
    <QuestionContent content={props.question.stem_html} source={props.source} />
  </QuestionWrapper>;
};

export default Question;
