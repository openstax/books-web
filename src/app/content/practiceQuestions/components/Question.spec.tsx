import React from 'react';
import { Provider } from 'react-redux';
import renderer, { act } from 'react-test-renderer';
import createTestServices from '../../../../test/createTestServices';
import createTestStore from '../../../../test/createTestStore';
import { book } from '../../../../test/mocks/archiveLoader';
import * as Services from '../../../context/Services';
import MessageProvider from '../../../MessageProvider';
import { Store } from '../../../types';
import { assertDefined } from '../../../utils';
import { LinkedArchiveTreeSection } from '../../types';
import { findArchiveTreeNodeById } from '../../utils/archiveTreeUtils';
import { nextQuestion, setAnswer, setQuestions, setSelectedSection } from '../actions';
import { PracticeQuestion } from '../types';
import Answer from './Answer';
import Question from './Question';

jest.mock('../../components/ContentExcerpt', () => (props: any) => <div data-mock-content-excerpt {...props} />);

describe('Question', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let render: () => JSX.Element;
  let linkedArchiveTreeSection: LinkedArchiveTreeSection;
  let dispatch: jest.SpyInstance;
  const mockQuestion = {
    answers: [
      {
        content_html: '<span data-math=\'25\'>25</span>',
        correctness: '0.0',
        feedback_html: '...',
        id: 273729,
      },
      {
        content_html: '<span data-math=\'26\'>26</span>',
        correctness: '1.0',
        feedback_html: 'Iron is most strongly bound nuclide.',
        id: 273730,
      },
    ],
    group_uuid: 'd95384f2-1330-4582-9d81-1af0eae17b48',
    stem_html: 'What is the atomic number of the most strongly bound nuclide?',
    tags: 'd95384f2-1330-4582-9d81-1af0eae17b48',
    uid: '11591@5',
  } as PracticeQuestion;

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
    dispatch = jest.spyOn(store, 'dispatch');
    render = () => <Provider store={store}>
      <Services.Provider value={services}>
        <MessageProvider>
          <Question />
        </MessageProvider>
      </Services.Provider>
    </Provider>;
    linkedArchiveTreeSection = assertDefined(
      findArchiveTreeNodeById(book.tree, 'testbook1-testpage2-uuid'),
      'mock file has been changed'
    ) as LinkedArchiveTreeSection;
  });

  it('renders null if there is no current question', () => {
    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
    const component = renderer.create(render());

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders null if there is no selected section question', () => {
    store.dispatch(setQuestions([mockQuestion]));
    store.dispatch(nextQuestion());
    const component = renderer.create(render());

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders properly', () => {
    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
    store.dispatch(setQuestions([mockQuestion]));
    store.dispatch(nextQuestion());

    const component = renderer.create(render());

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('submit is active after selecting an answer', () => {
    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
    store.dispatch(setQuestions([mockQuestion]));
    store.dispatch(nextQuestion());

    const component = renderer.create(render());

    // Run initial useEffect hook
    // tslint:disable-next-line: no-empty
    act(() => {});

    const submit = component.root.findByProps({ value: 'Submit' });
    expect(submit.props.disabled).toEqual(true);

    const [firstAnswer] = component.root.findAllByType(Answer);
    const input = firstAnswer.findByProps({ type: 'radio' });

    act(() => {
      input.props.onChange();
    });

    expect(submit.props.disabled).toEqual(false);
  });

  it('clicking skip works', () => {
    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
    store.dispatch(setQuestions([mockQuestion]));
    store.dispatch(nextQuestion());
    dispatch.mockClear();

    const component = renderer.create(render());

    // Run initial useEffect hook
    // tslint:disable-next-line: no-empty
    act(() => {});

    const skip = component.root.findByProps({ value: 'Skip' });

    act(() => {
      skip.props.onClick();
    });

    expect(dispatch).toHaveBeenCalledWith(setAnswer({ questionId: mockQuestion.uid, answer: null }));
    expect(dispatch).toHaveBeenCalledWith(nextQuestion());
  });

  it('after submitting incorrect answer there is Show answer button that works', () => {
    store.dispatch(setSelectedSection(linkedArchiveTreeSection));
    store.dispatch(setQuestions([mockQuestion]));
    store.dispatch(nextQuestion());
    dispatch.mockClear();

    const component = renderer.create(render());

    // Run initial useEffect hook
    // tslint:disable-next-line: no-empty
    act(() => {});

    const submit = component.root.findByProps({ value: 'Submit' });
    expect(submit.props.disabled).toEqual(true);

    const [firstAnswer, secondAnswer] = component.root.findAllByType(Answer);
    const input = firstAnswer.findByProps({ type: 'radio' });

    act(() => {
      input.props.onChange();
    });

    const form = component.root.findByProps({ 'data-testid': 'question-form' });
    const preventDefault = jest.fn();
    form.props.onSubmit({ preventDefault });

    expect(preventDefault).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith(setAnswer({ questionId: mockQuestion.uid, answer: mockQuestion.answers[0] }));
    expect(() => firstAnswer.findByProps({ id: 'i18n:practice-questions:popup:incorrect' })).not.toThrow();
    expect(() => secondAnswer.findByProps({ id: 'i18n:practice-questions:popup:correct' })).toThrow();

    const showAnswer = component.root.findByProps({ value: 'Show answer' })!;

    act(() => {
      showAnswer.props.onClick();
    });

    expect(() => secondAnswer.findByProps({ id: 'i18n:practice-questions:popup:correct' })).not.toThrow();
    expect(() => component.root.findByProps({ value: 'Show answer' })).toThrow();
  });
});
