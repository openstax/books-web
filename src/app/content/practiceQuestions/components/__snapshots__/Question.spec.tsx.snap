// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question renders null if there is no current question 1`] = `null`;

exports[`Question renders null if there is no selected section question 1`] = `null`;

exports[`Question renders properly 1`] = `
<form
  className="Question__QuestionWrapper-vvov7z-0 cUpLsP"
  data-testid="question-form"
  onSubmit={[Function]}
>
  <div
    className="Question__QuestionContent-vvov7z-1 LbXMr"
    content="What is the atomic number of the most strongly bound nuclide?"
    data-mock-content-excerpt={true}
    source={
      Object {
        "id": "testbook1-testpage2-uuid",
        "parent": Object {
          "contents": Array [
            Object {
              "id": "testbook1-testpage2-uuid@1.0",
              "slug": "1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units",
              "title": "<span class=\\"os-text\\">Introduction to Science and the Realm of Physics, Physical Quantities, and Units</span>",
            },
          ],
          "id": "testbook1-testchapter10@1.0",
          "parent": Object {
            "contents": Array [
              Object {
                "contents": Array [
                  Object {
                    "id": "testbook1-testpage2-uuid@1.0",
                    "slug": "1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units",
                    "title": "<span class=\\"os-text\\">Introduction to Science and the Realm of Physics, Physical Quantities, and Units</span>",
                  },
                ],
                "id": "testbook1-testchapter10@1.0",
                "slug": "testchapter10-slug",
                "title": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1.1</span>",
              },
              Object {
                "id": "testbook1-testpage11-uuid@1.0",
                "slug": "1-1-physics-an-introduction",
                "title": "<span class=\\"os-number\\">1.1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Physics: An Introduction</span></span>",
              },
              Object {
                "id": "testbook1-testpage8-uuid@1.0",
                "slug": "23-12-rlc-series-ac-circuits",
                "title": "<span class=\\"os-number\\">23.12</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">RLC Series AC Circuits</span></span>",
              },
            ],
            "id": "testbook1-testchapter1-uuid@1.0",
            "parent": Object {
              "contents": Array [
                Object {
                  "id": "testbook1-testpage1-uuid@1.0",
                  "slug": "test-page-1",
                  "title": "<span class=\\"os-text\\">Test Page 1</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "contents": Array [
                        Object {
                          "id": "testbook1-testpage2-uuid@1.0",
                          "slug": "1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units",
                          "title": "<span class=\\"os-text\\">Introduction to Science and the Realm of Physics, Physical Quantities, and Units</span>",
                        },
                      ],
                      "id": "testbook1-testchapter10@1.0",
                      "slug": "testchapter10-slug",
                      "title": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1.1</span>",
                    },
                    Object {
                      "id": "testbook1-testpage11-uuid@1.0",
                      "slug": "1-1-physics-an-introduction",
                      "title": "<span class=\\"os-number\\">1.1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Physics: An Introduction</span></span>",
                    },
                    Object {
                      "id": "testbook1-testpage8-uuid@1.0",
                      "slug": "23-12-rlc-series-ac-circuits",
                      "title": "<span class=\\"os-number\\">23.12</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">RLC Series AC Circuits</span></span>",
                    },
                  ],
                  "id": "testbook1-testchapter1-uuid@1.0",
                  "slug": "testchapter1-slug",
                  "title": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "id": "testbook1-testpage3-uuid@1.0",
                      "slug": "2-test-page-3",
                      "title": "<span class=\\"os-text\\">Test Page 3</span>",
                    },
                  ],
                  "id": "testbook1-testchapter2-uuid@1.0",
                  "slug": "testchapter2-slug",
                  "title": "<span class=\\"os-number\\">2</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 2</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "id": "testbook1-testpage4-uuid@1.0",
                      "slug": "3-test-page-4",
                      "title": "<span class=\\"os-text\\">Test Page 4</span>",
                    },
                  ],
                  "id": "testbook1-testchapter3-uuid@1.0",
                  "slug": "testchapter3-slug",
                  "title": "<span class=\\"os-number\\">3</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 3</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "id": "testbook1-testpage5-uuid@1.0",
                      "slug": "4-test-page-5",
                      "title": "<span class=\\"os-text\\">Test Page 5</span>",
                    },
                  ],
                  "id": "testbook1-testchapter4-uuid@1.0",
                  "slug": "testchapter4-slug",
                  "title": "<span class=\\"os-number\\">4</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 4</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "id": "testbook1-testpage6-uuid@1.0",
                      "slug": "10-test-page-6",
                      "title": "<span class=\\"os-text\\">Test Page 6</span>",
                    },
                  ],
                  "id": "testbook1-testchapter5-uuid@1.0",
                  "slug": "testchapter5-slug",
                  "title": "<span class=\\"os-number\\">10</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 5</span>",
                },
                Object {
                  "contents": Array [
                    Object {
                      "id": "testbook1-testpage7-uuid@1.0",
                      "slug": "12-test-page-7",
                      "title": "<span class=\\"os-text\\">Test Page 7</span>",
                    },
                  ],
                  "id": "testbook1-testchapter6-uuid@1.0",
                  "slug": "testchapter6-slug",
                  "title": "<span class=\\"os-number\\">12</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 6</span>",
                },
                Object {
                  "id": "testbook1-testpage9-uuid@1.0",
                  "slug": "a-atomic-masses",
                  "title": "<span class=\\"os-number\\">A</span><span class=\\"os-divider\\"> | </span><span class=\\"os-text\\">Atomic Masses</span>",
                },
                Object {
                  "id": "testbook1-testpage10-uuid@1.0",
                  "slug": "d-glossary-of-key-symbols-and-notation",
                  "title": "<span class=\\"os-number\\">D</span><span class=\\"os-divider\\"> | </span><span class=\\"os-text\\">Glossary of Key Symbols and Notation</span>",
                },
                Object {
                  "id": "testbook1-testpage12-uuid@1.0",
                  "slug": "test-page-for-generic-styles",
                  "title": "<span class=\\"os-text\\">Test Page for Generic Styles</span>",
                },
              ],
              "id": "testbook1-uuid@1.0",
              "slug": "testbook1-slug",
              "title": "Test Book 1",
            },
            "slug": "testchapter1-slug",
            "title": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1</span>",
          },
          "slug": "testchapter10-slug",
          "title": "<span class=\\"os-number\\">1</span><span class=\\"os-divider\\"> </span><span class=\\"os-text\\">Test Chapter 1.1</span>",
        },
        "slug": "1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units",
        "title": "<span class=\\"os-text\\">Introduction to Science and the Realm of Physics, Physical Quantities, and Units</span>",
        "version": undefined,
      }
    }
    tabIndex={0}
  />
  <div
    className="Question__AnswersWrapper-vvov7z-2 hNPliY"
  >
    <div
      className="Answer__AnswerBlock-n9titu-4 evBqvi"
      onClick={[Function]}
      style={
        Object {
          "focused": "#fff",
          "fontColor": "#606163",
          "hovered": "#0dc0dc",
          "label": "unselected",
          "passive": "#C6C6C6",
        }
      }
    >
      <label
        aria-label="Choice A"
        className="Answer__AnswerLabel-n9titu-3 gieTOT"
        style={
          Object {
            "focused": "#fff",
            "fontColor": "#606163",
            "hovered": "#0dc0dc",
            "label": "unselected",
            "passive": "#C6C6C6",
          }
        }
      >
        a
        <input
          checked={false}
          disabled={false}
          name="a"
          onChange={[Function]}
          type="radio"
        />
      </label>
      <div
        className="Answer__AnswerAlignment-n9titu-1 kPLngI"
      >
        <div
          className="Answer__AnswerContent-n9titu-2 dcoaiH"
        >
          <span
            className="Answer__AnswerExcerpt-n9titu-0 ecRPCq"
          >
            &lt;span data-math='25'&gt;25&lt;/span&gt;
          </span>
        </div>
      </div>
    </div>
    <div
      className="Answer__AnswerBlock-n9titu-4 evBqvi"
      onClick={[Function]}
      style={
        Object {
          "focused": "#fff",
          "fontColor": "#606163",
          "hovered": "#0dc0dc",
          "label": "unselected",
          "passive": "#C6C6C6",
        }
      }
    >
      <label
        aria-label="Choice B"
        className="Answer__AnswerLabel-n9titu-3 gieTOT"
        style={
          Object {
            "focused": "#fff",
            "fontColor": "#606163",
            "hovered": "#0dc0dc",
            "label": "unselected",
            "passive": "#C6C6C6",
          }
        }
      >
        b
        <input
          checked={false}
          disabled={false}
          name="b"
          onChange={[Function]}
          type="radio"
        />
      </label>
      <div
        className="Answer__AnswerAlignment-n9titu-1 kPLngI"
      >
        <div
          className="Answer__AnswerContent-n9titu-2 dcoaiH"
        >
          <span
            className="Answer__AnswerExcerpt-n9titu-0 ecRPCq"
          >
            &lt;span data-math='26'&gt;26&lt;/span&gt;
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="QuestionNavigation__Wrapper-sc-32mlh3-0 haPgDl"
  >
    <input
      className="PQButton__PQInput-ypamof-1 bnCeG"
      onClick={[Function]}
      type="button"
      value="Skip"
    />
    <input
      className="PQButton__PQInput-ypamof-1 fapUor"
      default={true}
      disabled={true}
      type="submit"
      value="Submit"
    />
  </div>
</form>
`;
